contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,188 @@
+            "Invalid Source Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(senderAddress == sourceContractAddress, "Unauthorized Sender");
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(recipient != address(0), "Invalid Recipient Address");
+
+        uint256 adjustedAmount = amount;
+        if (tokenReserve > 0) {
+            if (amount > tokenReserve) {
+                emit CollateralAdded({amount: tokenReserve, remaining: 0});
+                adjustedAmount = amount - tokenReserve;
+                tokenReserve = 0;
+            } else {
+                tokenReserve -= amount;
+                emit CollateralAdded({amount: amount, remaining: tokenReserve});
+                return;
+            }
+        }
+
+        // Calls NativeMinter precompile through INativeMinter interface.
+        _nativeMinter.mintNativeCoin(recipient, adjustedAmount);
+        emit MintNativeTokens(recipient, adjustedAmount);
+    }
+
+    /**
+     * @dev See {INativeTokenMinter-bridgeTokens}.
+     */
+    function transferToSource(
+        address recipient,
+        address feeContractAddress,
+        uint256 feeAmount
+    ) external payable nonReentrant {
+        // The recipient cannot be the zero address.
+        require(recipient != address(0), "Invalid Recipient Address");
+
+        require(
+            tokenReserve == 0,
+            "Cannot release tokens until source contract is collateralized"
+        );
+
+        // Lock tokens in this bridge instance. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedAmount = 0;
+        if (feeAmount > 0) {
+            adjustedAmount = SafeERC20TransferFrom.safeTransferFrom(
+                IERC20(feeContractAddress),
+                feeAmount
+            );
+            IERC20(feeContractAddress).safeIncreaseAllowance(
+                address(teleporterMessenger),
+                adjustedAmount
+            );
+        }
+
+        // Burn native token by sending to BLACKHOLE_ADDRESS
+        payable(BLACKHOLE_ADDRESS).transfer(msg.value);
+
+        // Send Teleporter message.
+        bytes memory messageData = abi.encode(recipient, msg.value);
+
+        uint256 messageID = teleporterMessenger.sendCrossChainMessage(
+            TeleporterMessageInput({
+                destinationChainID: sourceBlockchainID,
+                destinationAddress: sourceContractAddress,
+                feeInfo: TeleporterFeeInfo({
+                    contractAddress: feeContractAddress,
+                    amount: adjustedAmount
+                }),
+                requiredGasLimit: TRANSFER_NATIVE_TOKENS_REQUIRED_GAS,
+                allowedRelayerAddresses: new address[](0),
```

Comment:
I could imagine subnet operators might want to restrict who could relay these cross chain messages, as they are quite sensitive. Adding `allowedRelayerAddresses` as a `transferToSource` function parameter would make sense. Same comment applies for `transferToDestination` to `NativeTokenSource.sol`